tool/jt.rb

```diff
@@ -1050,85 +1051,98 @@ def test_compiler(*args)
-      out, _ = mx_sulong('lli', "-Dpolyglot.llvm.libraries=#{openssl_lib}", 'test/truffle/cexts/xopenssl/main.bc', capture: true)
-      raise out.inspect unless out == "5d41402abc4b2a76b9719d911017c592\n"
-    end
+            tests = [
+                ['oily_png', ['chunky_png-1.3.6', 'oily_png-1.2.0'], ['oily_png']],
+                ['psd_native', ['chunky_png-1.3.6', 'oily_png-1.2.0', 'bindata-2.3.1', 'hashie-3.4.4', 'psd-enginedata-1.1.1', 'psd-2.1.2', 'psd_native-1.1.3'], ['oily_png', 'psd_native']],
+                ['nokogiri', [], ['nokogiri']]
+            ]
 
-    # Test that we can compile and run some very basic C extensions
+            tests.each do |gem_name, dependencies, libs|
+              puts "", gem_name
+              next if gem_name == 'nokogiri' # nokogiri totally excluded
+              gem_root = "#{TRUFFLERUBY_DIR}/test/truffle/cexts/#{gem_name}"
+              ext_dir = Dir.glob("#{gem_home}/gems/#{gem_name}*/")[0] + "ext/#{gem_name}"
+
+              compile_cext gem_name, ext_dir, "#{gem_root}/lib/#{gem_name}/#{gem_name}.su", '-Werror=implicit-function-declaration'
+
+              next if gem_name == 'psd_native' # psd_native is excluded just for running
+              run_ruby *dependencies.map { |d| "-I#{gem_home}/gems/#{d}/lib" },
+                       *libs.map { |l| "-I#{TRUFFLERUBY_DIR}/test/truffle/cexts/#{l}/lib" },
+                       "#{TRUFFLERUBY_DIR}/test/truffle/cexts/#{gem_name}/test.rb", gem_root
+            end
+
+            # Tests using gem install to compile the cexts
+            sh "test/truffle/cexts/puma/puma.sh"
+            sh "test/truffle/cexts/sqlite3/sqlite3.sh"
+            sh "test/truffle/cexts/unf_ext/unf_ext.sh"
 
-    begin
-      output_file = 'cext-output.txt'
-      ['minimum', 'method', 'module', 'globals', 'backtraces', 'xopenssl'].each do |gem_name|
-        next if gem_name == 'xopenssl' && no_openssl
-        dir = "#{TRUFFLERUBY_DIR}/test/truffle/cexts/#{gem_name}"
-        ext_dir = "#{dir}/ext/#{gem_name}/"
-        compile_cext gem_name, ext_dir, "#{dir}/lib/#{gem_name}/#{gem_name}.su"
-        case gem_name
-        when 'backtraces'
-          run_ruby "-I#{dir}/lib", "#{dir}/bin/#{gem_name}", err: output_file, continue_on_failure: true
-          actual = File.read(output_file).gsub(TRUFFLERUBY_DIR, '').gsub(/\/cext(_ruby)?\.rb:(\d+)/, '/cext\1.rb:n')
-          expected = File.read("#{dir}/expected.txt")
-          unless actual == expected
-            abort "C extension #{dir} didn't work as expected\nActual:\n#{actual}\nExpected:\n#{expected}"
-          end
         else
-          run_ruby "-I#{dir}/lib", "#{dir}/bin/#{gem_name}", out: output_file
-          unless File.read(output_file) == File.read("#{dir}/expected.txt")
-            abort "c extension #{dir} didn't work as expected"
-          end
+          raise "unknown test: #{test_name}"
         end
       end
     ensure
       File.delete output_file rescue nil
```

Comment:
Is `output_file` set in this scope? It should be in the `when` above I believe.
And `rescue nil` => `if File.exist?(output_file)` would be better to notice these problems earlier.